src/Features/LanguageServer/Protocol/Handler/ProjectContext/GetTextDocumentWithContextHandler.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [Shared]
+    [ExportLspMethod(MSLSPMethods.ProjectContextsName)]
+    internal class GetTextDocumentWithContextHandler : IRequestHandler<GetTextDocumentWithContextParams, ActiveProjectContexts?>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public GetTextDocumentWithContextHandler()
+        {
+        }
+
+        public Task<ActiveProjectContexts?> HandleRequestAsync(
+            Solution solution,
+            GetTextDocumentWithContextParams request,
+            ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            var documents = solution.GetDocuments(request.TextDocument.Uri, clientName);
+
+            if (!documents.Any())
+            {
+                // TODO: am I supposed to return an empty list of contexts or null here?
+                return Task.FromResult<ActiveProjectContexts?>(null);
+            }
+
+            var contexts = new List<ProjectContext>();
+
+            foreach (var document in documents)
+            {
+                var project = document.Project;
+
+                contexts.Add(new ProjectContext
+                {
+                    Id = project.Id.Id.ToString(),
+                    Kind = project.Language == LanguageNames.CSharp ? ProjectContextKind.CSharp : ProjectContextKind.VisualBasic,
+                    Label = project.Name
+                });
+            }
+
+            // If the document is open, it doesn't matter which DocumentId we pass to GetDocumentIdInCurrentContext since
+            // all the documents are linked at that point, so we can just pass the first arbitrarily. If the document is closed
+            // GetDocumentIdInCurrentContext will just return the same ID back, which means we're going to pick the first
+            // ID in GetDocumentIdsWithFilePath, but there's really nothing we can do since we don't have contexts for
```

Comment:
will this lsp method get called for closed documents?  is that something we need to worry about?

Reply:
I don't know how it would get queried given usually the LSP is only ever querying about open files. Default context for a closed file isn't something we even really can answer today -- if the only contexts a file is in is a multi-targeting project we could figure that out. But linked files across projects there's no concept in VS we could query even if we wanted to...